src/MiniProfiler.AspNetCore.Mvc/MvcDiagnosticListener.cs

```diff
@@ -0,0 +1,166 @@
+            AfterActionFilterOnActionExecutedEventData data => Complete(data.Filter),
+
+            // ResultFilterOnResultExecution
+            BeforeResultFilterOnResultExecutionEventData data => Start(data.Filter, "Result Filter (Exec): " + GetName(data.Filter)),
+            AfterResultFilterOnResultExecutionEventData data => Complete(data.Filter),
+            // ResultFilterOnResultExecuting
+            BeforeResultFilterOnResultExecutingEventData data => Start(data.Filter, "Result Filter (Execing): " + GetName(data.Filter)),
+            AfterResultFilterOnResultExecutingEventData data => Complete(data.Filter),
+            // ResultFilterOnResultExecuted
+            BeforeResultFilterOnResultExecutedEventData data => Start(data.Filter, "Result Filter (Execed): " + GetName(data.Filter)),
+            AfterResultFilterOnResultExecutedEventData data => Complete(data.Filter),
+
+            // Razor Bits: https://github.com/dotnet/aspnetcore/blob/v3.0.0/src/Mvc/Mvc.Razor/src/Diagnostics/MvcDiagnostics.cs
+            // ViewPage
+            BeforeViewPageEventData data => Start(data.Page, "View: " + data.Page.Path),
+            AfterViewPageEventData data => Complete(data.Page),
+
+            // RazorPage Bits: https://github.com/dotnet/aspnetcore/blob/v3.0.0/src/Mvc/Mvc.RazorPages/src/Diagnostics/MvcDiagnostics.cs
+            // HandlerMethod
+            BeforeHandlerMethodEventData data => Start(data.Instance, "Handler: " + data.HandlerMethodDescriptor.Name),
+            AfterHandlerMethodEventData data => Complete(data.Instance),
+
+            // PageFilterOnPageHandlerExecution
+            BeforePageFilterOnPageHandlerExecutionEventData data => Start(data.Filter, "Filter (Exec): " + GetName(data.Filter)),
+            AfterPageFilterOnPageHandlerExecutionEventData data => Complete(data.Filter),
+            // PageFilterOnPageHandlerExecuting
+            BeforePageFilterOnPageHandlerExecutingEventData data => Start(data.Filter, "Filter (Execing): " + GetName(data.Filter)),
+            AfterPageFilterOnPageHandlerExecutingEventData data => Complete(data.Filter),
+            // PageFilterOnPageHandlerExecuted
+            BeforePageFilterOnPageHandlerExecutedEventData data => Start(data.Filter, "Filter (Execed): " + GetName(data.Filter)),
+            AfterPageFilterOnPageHandlerExecutedEventData data => Complete(data.Filter),
+
+            // PageFilterOnPageHandlerSelection
+            BeforePageFilterOnPageHandlerSelectionEventData data => Start(data.Filter, "Filter (Selection): " + GetName(data.Filter)),
+            AfterPageFilterOnPageHandlerSelectionEventData data => Complete(data.Filter),
+            // PageFilterOnPageHandlerSelected
+            BeforePageFilterOnPageHandlerSelectedEventData data => Start(data.Filter, "Filter (Selected): " + GetName(data.Filter)),
+            AfterPageFilterOnPageHandlerSelectedEventData data => Complete(data.Filter),
```

Comment:
How badly will things get messed up if the *end* event never fires due to an exception?

I'm not sure that we provide the guarantee that both events fire when an exception happens.

Reply:
When a request ends, we go through the tree and stop all running timers (it'd be the head), so no big deal. The timing may be inaccurate but...I can't imagine you'd care it's off really. If it's the best we can do, doubly fine with that - thoughts?

Reply:
Seems fine, I just wanted to make sure you were aware of the possibility.